sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala

```diff
@@ -856,7 +857,137 @@ trait GpuAggregateWindowFunction[T <: Aggregation with RollingAggregation[T]]
+   *                          it can change for different rows than a column vector is provided.
+   *                          Only values that are for the same partition by keys should be
+   *                          modified. Because the input data is sorted by the partition by
+   *                          columns the boolean values will be grouped together.
+   * @param windowedColumnOutput the output of the windowAggregation without anything
+   *                             fixed/modified.
+   * @return a fixed ColumnVector that was with outputs updated for items that were in the same
+   *         group by key as the last row in the previous batch.
+   */
+  def fixUp(
+      samePartitionMask: Either[GpuColumnVector, Boolean],
+      windowedColumnOutput: GpuColumnVector): GpuColumnVector
+}
+
+/**
+ * For many operations a running window (unbounded preceding to current row) can
+ * process the data without dividing the data up into batches that contain all of the data
+ * for a given group by key set. Instead we store a small amount of state from a previous result
+ * and use it to fix the final result. This is a memory optimization. Any
+ * GpuAggregateWindowFunction can still work for running window queries. This just reduces the
+ * maximum amount of memory needed to process them.
+ */
+trait GpuBatchedRunningWindowFunction[T <: Aggregation with RollingAggregation[T]]
+    extends GpuAggregateWindowFunction[T] {
+
+  /**
+   * Get a new class that can be used to fix up batched RunningWindowOperations.
+   */
+  def newFixer(): BatchedRunningWindowFixer
+}
+
+/**
+ * This class fixes up batched running windows by performing a binary op on the previous value and
+ * those in the the same partition by key group. It does not deal with nulls, so it works for things
+ * like row_number and count, that cannot produce nulls, or for NULL_MIN and NULL_MAX that do the
+ * right thing when they see a null.
+ */
+class BatchedRunningWindowBinaryFixer(val binOp: BinaryOp, val name: String)
+    extends BatchedRunningWindowFixer with Arm with Logging {
+  private var previousResult: Option[Scalar] = None
+
+  override def updateState(finalOutputColumn: GpuColumnVector): Unit = {
+    logDebug(s"$name: updateState from $previousResult to...")
+    previousResult.foreach(_.close)
+    previousResult =
+      Some(finalOutputColumn.getBase.getScalarElement(finalOutputColumn.getRowCount.toInt - 1))
+    logDebug(s"$name: ... $previousResult")
+  }
+
+  override def fixUp(samePartitionMask: Either[GpuColumnVector, Boolean],
+      windowedColumnOutput: GpuColumnVector): GpuColumnVector = {
+    logDebug(s"$name: fix up $previousResult $samePartitionMask")
+    (previousResult, samePartitionMask) match {
+      case (None, _) => windowedColumnOutput.incRefCount()
+      case (Some(prev), scala.util.Right(mask)) =>
+        if (mask) {
+          GpuColumnVector.from(windowedColumnOutput.getBase.binaryOp(binOp, prev, prev.getType),
```

Comment:
so if mask == true, this is creating a new column vector from the result of `binaryOp`, should it consume `windowedColumnOutput`?

Reply:
No I will update the docs.